affinity/cpp-23/d1795r1.md

```diff
@@ -124,11 +132,135 @@ As a result of the above this paper may also:
 * Propose a lifetime model for execution agents.
 * Propose some additions to the C\+\+ machine model to facilitate describing these additional properties.
 
-# Suggested straw polls
+# 6. Proposal
+
+## Header `<system>` synopsis
+
+```cpp
+namespace std {
+namespace experimental {
+
+/* system_topology */
+
+class system_topology {
+
+  system_topology() = delete;
+
+  std::chrono::time_point<std::chrono::system_clock> timestamp() const noexcept;
+
+};
+
+/* system_resource */
+
+class system_resource {
+
+  system_resource() = delete;
+
+};
+
+/* traverse_topology */
+
+template <class T>
+std::vector<system_resource> traverse_topology(const system_topology &, const T &) noexcept;
+
+/* this_system::discover_topology */
+
+namespace this_system {
+
+system_topology discover_topology();
+
+}  // namespace this_system
+
+}  // experimental
+}  // std
+```
+
+## Terms of art
+
+The term *system resource* refers to a hardware or software abstraction of an execution, memory, network or I/O resource within a system.
+
+The term *system topology* refers to a possibly cyclic graph of *execution resources* connected to the abstract machine, and their various properties.
+
+> [*Note:* The current definition of *system topology* is currently incomplete and will be developed over the course of this proposal as the various C\+\+ domains are represented. *--end note*]
+
+The term *topology traversal policy* refers to a policy that describes the way in which a *system topology* is traversed in order to to produce a collection of *system resources*.
+
+## Class `system_topology`
+
+The `system_topology` class provides an abstraction of a read-only snapshot of the *system topology* at a particular point in time. A `system_topology` object may not maintain or otherwise be associated with the lifetime of operating system or third party library resources.
+
+### `system_topology` constructors
+
+```cpp
+system_topology() = delete;
+```
+
+*Effects:* Explicitly deleted.
+
+### `timestamp` member function
+
+```cpp
+std::chrono::time_point<std::chrono::system_clock> timestamp() const noexcept;
+```
+
+*Returns:* A `std::chrono::time_point<std::chrono::system_clock>` object representing the time at which the runtime discovery of the system topology performed to construct the `system_topology` object was completed.
+
+*Throws:* May not throw.
+
+## Class `system_resource`
+
+The `system_resource` class provides an abstraction of a read-only snapshot of a *system resource* from the *system topology* at a particular point in time. A `system_resource` object may not maintain or otherwise be associated with the lifetime of operating system or third party library resources.
+
+### `system_resource` constructors
+
+```cpp
+system_resource() = delete;
```

Comment:
That's a good point, thanks for catching that! I believe we ran into the same issue in an earlier revision of the paper.

Though now that C++20 has ranges we can have this return a `ranges::view`, this also allows us to do range adaptations on top for further filtering the topology results.

Reply:
We did have a debate earlier about whether getting the topology should return a copy (`vector`) or a view (`span` etc.).  Both have issues: the former implies semiregularity of `system_resource` (and thus implies either reference counting or some way of assigning unique IDs), while the latter constrains the topology's lifetime in ways we don't want.

Reply:
You make a good point here. I can see value in both of these approaches, I think having lazy views over the `system_resource`s within a `system_topology` could be a nice way to compose topology traversal routines, but I also agree that it's important that a `system_resource` has a unique identifier for the reasons you mentioned and that we ultimately want to be able to store the resulting `system_resource`(s) without worrying about their lifetimes being tied to the `system_topology` object.

Perhaps we could have the best of both here, we could have the `system_resource` be `semiregular` and provide a unique identifier, and then have `traverse_topology` return a `ranges::view` that is temporarily tied to the lifetime of the `system_topology` object but capable of being assigned to a `vector` storing a copy of each resulting `system_resource`. I believe this should work, though I am not an expert in ranges and will have to look into this some more.

Though I also think this is tied to some interesting questions about the lifetime of the system topology and the objects associated with it (`system_resource`s, but also at some point `executor`s, `allocator`s, etc). I believe we made a lot of progress in this area in P0796 when defining the semantics of the `execution_resource`, so we should revisit that wording and look at incorporating that into the new paper for the next revision.

For now, I will leave the specific type returned from `traverse_topology` as to be decided with a note describing the two options.

Reply:
All true.  For me, I'd lean toward the view option with a lifetime tied to the system_topology if only because I can't think of a way to work with the result if it can be randomly invalidated asynchronously with my code and would prefer not to incur an extra copy to get that safety.